keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,965 @@
+    for _, obj := range objects {
+        // This logger variable shadows the function parameter.
+        logger := logger.WithValue(“obj”, obj)
+        doSomething(logger, obj)
+    }
+}
+```
+
+#### Unused return values
+
+This code looks like it adds a name, but isn’t doing it correctly:
+
+```go
+func foo(logger klogr.Logger, objects ...string) {
+        // BUG: WithName returns a new logger with the name,
+        // but that return value is not used.
+        logger.WithName(“foo”)
+        doSomething(logger)
+    }
+}
+```
+
+
+## Design Details
+
+klog currently provides a formatter for log messages, a global logger instance,
+and some helper code.
+
+<<[UNRESOLVED @pohly ]>>
+[Log sanitization](https://github.com/kubernetes/enhancements/issues/1753) is an
+alpha feature with unclear next steps. Unless there is sufficient interest in
+that feature, it will not be supported anymore.
+<<[/UNRESOLVED]>>
+
+A new package `k8s.io/klogr` will replace both `k8s.io/klog` and
+`k8s.io/klog/klogr`. It will be hosted under `staging` and thus get developed
+together with the code that uses it, without a stable API guarantee. It’s
+content is described in the following sections.
+
+A new staging repo is most suitable because placing the package in some of the
+other potential places would create cyclic dependencies at the module level
+(`k8s.io/component-helper` -> `k8s.io/client-go` -> `k8s.io/component-helper`),
+something that `hack/update-vendor.sh` complains about, or force consumers to
+deal with an unnecessarily large module (`k8s.io/component-base`).
+
+### Text format
+
+The formatting code will be moved into a standalone `go-logr/logr.LogSink` in the
+`k8s.io/klogr/logger` package. The goal is to produce identical output to the current
+format, but with `logr.Logger` as interface and just the klog features that
+weren’t deprecated (i.e. `-v` and `-vmodule`).
+
+### Default logger
+
+This is meant to replace existing solutions like the global [`Log` instance in
+controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/blob/78ce10e2ebad9205eff8429c3f0556788d680c27/pkg/log/log.go#L77-L82). It
+is needed as fallback for functions where a logger:
+- cannot be passed into the function at all or
+- a context was passed without an attached logger.
+
+The goal is to make passing a logger possible everywhere, so the first case
+should not occur anymore at some point. The second case remains valid because
+it is conventiont to just initialize the global logger once and then have all
+code pick it up from there.
+
+The API is:
+
+```go
+// SetFallbackLogger may only be called once per program run. Setting it
+// multiple times would not have the intended effect because the new logger
+// would not be used by code which already retrieved the fallback logger
+// earlier.
+//
+// Therefore this also should be called before any code starts to use logging.
+// Such code will get a valid logger to avoid crashes, but that logger will add
```

Comment:
My idea: as well as prefixing, always log an internal error (that logging was used with no fallback logger set).

Reply:
That sounds like a good idea. It's more likely to show up on someone's radar than the info message itself. Will add it.